src/thor/expansion_action.cc

```diff
@@ -0,0 +1,176 @@
+#include "thor/worker.h"
+
+#include "baldr/json.h"
+#include "baldr/rapidjson_utils.h"
+#include "midgard/constants.h"
+#include "midgard/logging.h"
+#include "midgard/polyline2.h"
+#include "midgard/util.h"
+
+using namespace rapidjson;
+using namespace valhalla::midgard;
+
+namespace {
+using namespace valhalla;
+static std::unordered_map<Options::ExpansionProps, const char*> kPropPaths =
+    {{Options_ExpansionProps_edge_ids, "/features/0/properties/edge_ids"},
+     {Options_ExpansionProps_statuses, "/features/0/properties/statuses"},
+     {Options_ExpansionProps_distances, "/features/0/properties/distances"},
+     {Options_ExpansionProps_durations, "/features/0/properties/durations"},
+     {Options_ExpansionProps_costs, "/features/0/properties/costs"}};
+} // namespace
+
+namespace valhalla {
+namespace thor {
+
+std::string thor_worker_t::expansion(Api& request) {
+  // time this whole method and save that statistic
+  measure_scope_time(request, "thor_worker_t::expansion");
+
+  // get the request params
+  auto options = request.options();
+  auto exp_action = options.expansion_action();
+  bool skip_opps = options.skip_opposites();
+  auto exp_props = options.expansion_props();
+  std::unordered_set<baldr::GraphId> opp_edges;
+
+  // default generalization to ~ zoom level 15
+  float gen_factor = options.has_generalize() ? options.generalize() : 10.f;
```

Comment:
this generalization works fine.. had an earlier comment with some some stats and a pic for 10 and 20 m epsilon, 10m wasn't that much different in size but much better geometries. I guess the difference would be bigger in autobahn areas